### PR TITLE
TypeHints moved from Constraint Field to Variable Field

### PR DESCRIPTION
**Pull request by maxbaxt**
## What is the goal of this PR?

Moved the Type Hints (the set of possible types that a variable can be), from one of its constraints (`isaConstraint` for `ThingVariable` and `subConstraint` for `TypeVariable`) to `Variable`. `TypeHints` in `RolePlayer` remain unchanged.

We also introduce a new `isSatisfiable` flag onto `Variable`. This acts as a guarantee of the satisfiability of a `Variable`, and is important in the case of an empty `TypeHints` (which can mean either a Variable that can't be anything, or can be anything, depending upon this flag).

## What are the changes implemented in this PR?

Field `typeHints` and its accompanying methods were moved from `isaConstraint` and `subConstraint` to `Variable`. Code calling on typeHints have been changed accordingly.

Added `isSatisfiable` flag, and added some logic to changing it when TypeHints are added to and adjusted in `TypeHinter` and `Rule`.

### CODE REVIEW

File: pattern/variable/Variable.java [hunk 1]
```diff
@@ -52,6 +58,7 @@ public Reference reference() {
     }
 
     public void addTo(Traversal traversal) {
+        if (!typeHints.isEmpty()) traversal.types(identifier, typeHints);
```
Comment (maxbaxt) [inline pattern/variable/Variable.java:61]: Don't we need to add the isSatisfiable information to the vertex as well?

File: pattern/variable/Variable.java [hunk 2]
```diff
@@ -71,6 +78,34 @@ public ThingVariable asThing() {
...
+            return;
+        }
+        typeHints.retainAll(labels);
+        if (typeHints.isEmpty()) isSatisfiable = false;
+    }
+
+    public void removeHint(Label label) {
+        assert !(typeHints.isEmpty() && isSatisfiable);
+        typeHints.remove(label);
+    }
```
Comment (maxbaxt) [inline pattern/variable/Variable.java:99]: removeHint() is only used when removing types that don't conform to being a sub of a hint of the variables sub constraint. When the Variable has no typeHints, this method is not used, and instead all the subTypes of it's sub constraint is inserted as its TypeHints.

File: logic/Rule.java [hunk 3]
```diff
@@ -162,19 +160,12 @@ void validateCycles() {
...
-                        whenVar.get().asType().sub().ifPresent(whenSub -> thenVar.asType().sub().ifPresent(
-                                thenSub -> thenSub.retainHints(whenSub.getTypeHints())));
-                    }
-                });
+        then.variables().stream().filter(variable -> variable.identifier().isNamedReference())
+                .forEach(thenVar ->
+                        when.variables().stream()
+                                .filter(variable -> variable.identifier().equals(thenVar.identifier()))
+                                .findFirst().ifPresent(whenVar -> thenVar.retainHints(whenVar.typeHints()))
+                );
```
Comment (maxbaxt) [inline logic/Rule.java:168]: Do tests exist for this method to ensure it still works as before?

File: logic/concludable/ConjunctionConcludable.java [hunk 4]
```diff
@@ -170,9 +169,6 @@ public Isa(final IsaConstraint constraint) {
 
         @Override
         public Stream<Unification> unify(ThenConcludable.Isa headIsa) {
-            Set<Label> possibleLabels = constraint().getTypeHints();
-            Set<Label> otherLabels = headIsa.constraint().getTypeHints();
-            possibleLabels.retainAll(otherLabels);
             return Stream.of(new Unification(this, headIsa, null)); // TODO Add variable mapping
```
Comment (flyingsilverfin) [inline logic/concludable/ConjunctionConcludable.java:172]: this is properly repalced by your other PR right?

File: pattern/variable/ThingVariable.java [hunk 5]
```diff
@@ -56,6 +57,7 @@
     private final Set<HasConstraint> hasConstraints;
     private final Set<ValueConstraint<?>> valueConstraints;
     private final Set<ThingConstraint> constraints;
+    private final Set<Label> typeHints;
```
Comment (flyingsilverfin) [inline pattern/variable/ThingVariable.java:60]: why is this here and on `Variable`?

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]

File: pattern/variable/ThingVariable.java (hunk 5 above)
Comment (maxbaxt) [reply pattern/variable/ThingVariable.java]: That shouldn't be there, deleted.

File: logic/concludable/ConjunctionConcludable.java (hunk 4 above)
Comment (maxbaxt) [reply logic/concludable/ConjunctionConcludable.java:172]: Indeed.

File: pattern/variable/Variable.java [hunk 6]
```diff
@@ -74,6 +81,34 @@ public ThingVariable asThing() {
         throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(ThingVariable.class));
     }
 
+    public void addHints(Set<Label> labels) {
```
Comment (flyingsilverfin) [inline pattern/variable/Variable.java:84]: I thought i had pointed this out before hmmm maybe I didn't hit submit:
Unless you can convince me otherwise, I think the API for adding/retaining/modifying type hints should _not_ be overlapping with the `isSatisfiable()` API internally. An external user can `setSatisfiable(false/true)` from outside, but from the variable's point of view, it is disconnected from type hints completely. This means the following methods should be simple adders/removers/setters, and the external user should use them to determine satisfiability (if using hints to do this at all!)

Comment (maxbaxt) [reply pattern/variable/Variable.java:84]: There was a misalignment with the use of `isSatisfiable` here. I've now moved the logic of changing `isSatisfiable` to outside of `Variable`, which means that we have to do these checks when some of these methods are called. In particular, I've made those checks on `TypeHinter`, and also on `Rule` where the `TypeHints` are pruned based on the `TypeHints` on the when conjunction. I suggest you confer that's the correct behaviour.

File: logic/Rule.java [hunk 7]
```diff
@@ -162,19 +161,16 @@ void validateCycles() {
...
-                    }
-                });
+        then.variables().stream().filter(variable -> variable.identifier().isNamedReference())
+                .forEach(thenVar ->
+                        when.variables().stream()
+                                .filter(whenVar -> whenVar.identifier().equals(thenVar.identifier()))
+                                .filter(whenVar -> !(whenVar.isSatisfiable() && whenVar.typeHints().isEmpty()))
+                                .findFirst().ifPresent(whenVar -> {
+                                    thenVar.retainHints(whenVar.typeHints());
+                                    if (thenVar.typeHints().isEmpty()) thenVar.setSatisfiability(false);
```
Comment (maxbaxt) [inline logic/Rule.java:171]: If the last of the TypeHints are pruned by the When then the variable needs to be set to Unsatisfiable.

File: logic/Rule.java [hunk 8]
```diff
@@ -162,19 +161,18 @@ void validateCycles() {
...
+        then.variables().stream().filter(variable -> variable.identifier().isNamedReference())
+                .forEach(thenVar ->
+                        when.variables().stream()
+                                .filter(whenVar -> whenVar.identifier().equals(thenVar.identifier()))
+                                .filter(whenVar -> !(whenVar.isSatisfiable() && whenVar.typeHints().isEmpty()))
+                                .findFirst().ifPresent(whenVar -> {
+                                    if (thenVar.typeHints().isEmpty() && thenVar.isSatisfiable()) {
+                                        thenVar.addHints(whenVar.typeHints());
+                                    } else thenVar.retainHints(whenVar.typeHints());
+                                    if (thenVar.typeHints().isEmpty()) thenVar.setSatisfiability(false);
```
Comment (maxbaxt) [inline logic/Rule.java:173]: If the last of the TypeHints are pruned by the When then the variable needs to be set to Unsatisfiable.

File: logic/tool/TypeHinter.java [hunk 9]
```diff
@@ -77,17 +77,17 @@ public Conjunction computeHintsExhaustive(Conjunction conjunction, int paralleli
         Map<Label, TypeVariable> labelMap = labelVarsFromConjunction(conjunction);
         Map<Reference, Set<Label>> referenceHintsMapping =
                 retrieveVariableHints(new HashSet<>(variableHints.getVariableHints()), parallelisation);
-        long numOfThings = traversalEng.graph().schema().stats().thingTypeCount();
+        long numOfTypes = traversalEng.graph().schema().stats().thingTypeCount();
```
Comment (maxbaxt) [inline logic/tool/TypeHinter.java:80]: Realised this name was inaccurate.

File: logic/tool/TypeHinter.java [hunk 10]
```diff
@@ -177,15 +153,15 @@ private void removeHintsViolatingSuper(Variable subVar) {
         if (supVar == null) return;
         if (supVar.reference().isLabel()) return;
 
-        Set<Label> supLabels = getHintLabels(supVar);
-        Set<Label> subLabels = getHintLabels(subVar);
-        if (supLabels == null || supLabels.isEmpty()) return;
+        Set<Label> supLabels = supVar.typeHints();
```
Comment (flyingsilverfin) [inline logic/tool/TypeHinter.java:156]: woop woop!

File: logic/tool/TypeHinter.java [hunk 11]
```diff
@@ -177,15 +153,15 @@ private void removeHintsViolatingSuper(Variable subVar) {
...
+        Set<Label> subLabels = subVar.typeHints();
+        if (supLabels.isEmpty()) return;
 
-        if (subLabels == null) return;
         if (subLabels.isEmpty()) {
             Set<Label> subHintsOfSupLabels = supLabels.stream().flatMap(label -> getType(label).getSubtypes())
                     .map(Type::getLabel).collect(Collectors.toSet());
-            addHintLabels(subVar, subHintsOfSupLabels);
+            subVar.addHints(subHintsOfSupLabels);
+            if (subVar.typeHints().isEmpty()) subVar.setSatisfiability(false);
```
Comment (flyingsilverfin) [inline logic/tool/TypeHinter.java:164]: how's it possible to add hints and its set to `false`??

File: pattern/constraint/thing/IsaConstraint.java [hunk 12]
```diff
@@ -63,31 +58,10 @@ public boolean isExplicit() {
...
-        typeHints.clear();
-    }
-
-    public Set<Label> getTypeHints() {
-        return typeHints;
-    }
-
     @Override
     public void addTo(Traversal traversal) {
         // TODO: assert !(type.reference().isLabel() && typeHints.isEmpty());
```
Comment (flyingsilverfin) [inline pattern/constraint/thing/IsaConstraint.java:63]: probably this assertion needs to move to wherever the type hints are now too?

File: logic/tool/TypeHinter.java (hunk 11 above)
Comment (maxbaxt) [reply logic/tool/TypeHinter.java:164]: `subHintsOfSupLabels` could be empty.

It's possible I might need to rename `addHintLabels` to `setHintLabels`

File: pattern/variable/Variable.java [hunk 13]
```diff
@@ -74,6 +81,31 @@ public ThingVariable asThing() {
...
+
+    public Set<Label> typeHints() {
+        return typeHints;
+    }
+
+    public boolean isSatisfiable() {
+        return isSatisfiable;
+    }
+
+    public void setSatisfiability(boolean isSatisfiable) {
```
Comment (flyingsilverfin) [inline pattern/variable/Variable.java:105]: i'd just call this `setSatisfiable` to only have 1 terminology

File: pattern/variable/Variable.java [hunk 14]
```diff
@@ -74,6 +81,31 @@ public ThingVariable asThing() {
...
+    public void addHints(Set<Label> labels) {
+        typeHints.addAll(labels);
+    }
+
+    public void retainHints(Set<Label> labels) {
+        typeHints.retainAll(labels);
+    }
+
+    public void removeHint(Label label) {
+        assert !(isSatisfiable && typeHints.isEmpty());
```
Comment (flyingsilverfin) [inline pattern/variable/Variable.java:93]: i think it'll be more readable with two negated ORs, but this is subjective :D

**Review by flyingsilverfin — APPROVED** [review]

File: pattern/variable/Variable.java (hunk 13 above)
Comment (maxbaxt) [reply pattern/variable/Variable.java:105]: Changed to `setIsSatisfiable`.

File: pattern/variable/Variable.java (hunk 14 above)
Comment (maxbaxt) [reply pattern/variable/Variable.java:93]: In general, I think `!(A ^ B)` and `(!A v !B)` are equally acceptable imho.

File: logic/tool/TypeHinter.java (hunk 11 above)
Comment (maxbaxt) [reply logic/tool/TypeHinter.java]: On second thoughts, we've already checked if supLabels is empty by this point, so I will remove this line.

File: pattern/constraint/thing/IsaConstraint.java (hunk 12 above)
Comment (flyingsilverfin) [reply pattern/constraint/thing/IsaConstraint.java:63]: @maxbaxt you did not remove this assertion........

Comment (maxbaxt) [reply pattern/constraint/thing/IsaConstraint.java:63]: You mean the comment? Deleted
